From 9f74550957ba2ac32f8fd17c2cbc7bd70f22650d Mon Sep 17 00:00:00 2001
From: srcnert <sercanerat@gmail.com>
Date: Sun, 15 Dec 2024 22:54:53 +0300
Subject: [PATCH] Fixing apollo3 adc problem

---
 drivers/adc/adc_ambiq.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/adc/adc_ambiq.c b/drivers/adc/adc_ambiq.c
index ce8ee479d16..3174c2bee9a 100644
--- a/drivers/adc/adc_ambiq.c
+++ b/drivers/adc/adc_ambiq.c
@@ -12,6 +12,8 @@
 #include <zephyr/pm/device_runtime.h>
 #include <zephyr/kernel.h>
 
+#define ADC_CONTEXT_WAIT_FOR_COMPLETION_TIMEOUT \
+	K_MSEC(1000)
 #define ADC_CONTEXT_USES_KERNEL_TIMER
 #include "adc_context.h"
 
@@ -271,6 +273,8 @@ static void adc_context_start_sampling(struct adc_context *ctx)
 	am_hal_adc_enable(data->adcHandle);
 	/*Trigger the ADC*/
 	am_hal_adc_sw_trigger(data->adcHandle);
+
+	k_busy_wait(1000);
 }
 
 static int adc_ambiq_init(const struct device *dev)
-- 
2.45.0

